pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -151,50 +313,89 @@ func NewDevicePluginHandlerImpl(updateCapacityFunc func(v1.ResourceList)) (*Devi
+	if devices != nil {
+		glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, contName, podUID, devices.List())
+		needed = needed - devices.Len()
+		// A pod's resource is not expected to change once admitted by the API server,
+		// so just fail loudly here. We can revisit this part if this no longer holds.
+		if needed != 0 {
+			return nil, fmt.Errorf("pod %v container %v changed request for resource %v from %v to %v", podUID, contName, resource, devices.Len(), required)
+		}
+	}
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
+	devices = sets.NewString()
+	// Needs to allocate additional devices.
+	if h.allocatedDevices[resource] == nil {
+		h.allocatedDevices[resource] = sets.NewString()
+	}
+	// Gets Devices in use.
+	devicesInUse := h.allocatedDevices[resource]
+	// Gets a list of available devices.
+	available := h.allDevices[resource].Difference(devicesInUse)
+	if int(available.Len()) < needed {
+		return nil, fmt.Errorf("requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, needed, available.Len())
+	}
+	allocated := available.UnsortedList()[:needed]
+	// Updates h.allocatedDevices with allocated devices to prevent them
+	// from being allocated to other pods/containers, given that we are
+	// not holding lock during the rpc call.
+	for _, device := range allocated {
+		h.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+	}
+	return devices, nil
+}
+
+// Allocate attempts to allocate all of required extended resources for
+// the input container, issues an Allocate rpc request for each of such
+// resources, processes their AllocateResponses, and updates the cached
+// containerDevices on success.
+func (h *HandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) error {
+	podUID := string(pod.UID)
+	contName := container.Name
+	allocatedDevicesUpdated := false
 	for k, v := range container.Resources.Limits {
 		resource := string(k)
 		needed := int(v.Value())
 		glog.V(3).Infof("needs %d %s", needed, resource)
-		_, registeredResource := h.allDevices[resource]
-		if !registeredResource || needed == 0 {
+		if _, registeredResource := h.allDevices[resource]; !registeredResource {
 			continue
 		}
-		h.Lock()
-		// Gets list of devices that have already been allocated.
-		// This can happen if a container restarts for example.
-		if h.allocatedDevices[resource] == nil {
-			h.allocatedDevices[resource] = make(podDevices)
-		}
-		devices := h.allocatedDevices[resource].getDevices(string(pod.UID), container.Name)
-		if devices != nil {
-			glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, container.Name, pod.UID, devices.List())
-			needed = needed - devices.Len()
+		// Updates allocatedDevices to garbage collect any stranded resources
+		// before doing the device plugin allocation.
```

Comment:
Question for the future: How do we expect to report used / unused devices ?

With the current model if devices get stranded they would be advertised as taken right ?
In that case wouldn't a model that cleans up stranded devices be a better one ?

Reply:
Good point, I added the code to restore h.allocatedDevices upon allocation failures.